lib/translate/inquiry.cc

```diff
@@ -0,0 +1,298 @@
+  // the Identify Controller Data Structure
+  memcpy(result.product_identification, identify_controller_data.mn, 16);
+
+  /*
+  Shall be set to the last 4 ASCII graphic characters in the range of 21h-7Eh
+  (i.e. last 4 non-space characters) of the Firmware Revision (FR) field within
+  the Identify Controller Data Structure.
+  */
+  int idx = 3;
+  for (int i = 7; i >= 0; i--) {
+    if (identify_controller_data.fr[i] != ' ' &&
+        identify_controller_data.fr[i] >= 0x21 &&
+        identify_controller_data.fr[i] <= 0x7e) {
+      result.product_revision_level[idx] = identify_controller_data.fr[i];
+      if (idx-- == 0) break;
+    }
+  }
+  if (idx >= 0) printf("less than four characters set\n");
+  return result;
+}
+scsi_defs::InquiryData build_standard_inquiry() {
+  // Identify controller results
+  nvme_defs::IdentifyControllerData identify_controller_data =
+      nvme_identify_controller();
+  // Identify namespace results
+  nvme_defs::IdentifyNamespace identify_namespace_data =
+      nvme_identify_namespace();
+  return translate_standard_inquiry_response(identify_controller_data,
+                                             identify_namespace_data);
+}
+
+scsi_defs::SupportedVitalProductData build_supported_vpd_pages() {
+  scsi_defs::SupportedVitalProductData result =
+      scsi_defs::SupportedVitalProductData();
+  result.peripheral_qualifier =
+      scsi_defs::PeripheralQualifier::kPeripheralDeviceConnected;
+  result.peripheral_device_type =
+      scsi_defs::PeripheralDeviceType::kDirectAccessBlock;
+  result.page_code = 0;
+
+  // Shall be set to 5 indicating number of items supported VPD pages list
+  // requires. NOTE: document says to set this to 5 but there are 7 entries....
+  result.page_length = 5;
+
+  result.supported_page_list[0] = 0x00;
+  result.supported_page_list[1] = 0x80;
+  result.supported_page_list[2] = 0x83;
+  result.supported_page_list[3] = 0x86;
+  result.supported_page_list[4] = 0xB0;
+  result.supported_page_list[5] = 0xB1;
+  result.supported_page_list[6] = 0xB2;
+  return result;
+}
+
+scsi_defs::UnitSerialNumber translate_unit_serial_number_vpd_response(
+    nvme_defs::IdentifyNamespace identify_namespace_data) {
+  scsi_defs::UnitSerialNumber result = scsi_defs::UnitSerialNumber();
+  result.peripheral_qualifier =
+      scsi_defs::PeripheralQualifier::kPeripheralDeviceConnected;
+  result.peripheral_device_type =
+      scsi_defs::PeripheralDeviceType::kDirectAccessBlock;
+  result.page_code = 0x80;
+
+  // check if nonzero
+  bool nguid_nz =
+      identify_namespace_data.nguid[0] || identify_namespace_data.nguid[1];
+  bool eui64_nz = identify_namespace_data.eui64;
+  if (eui64_nz) {
+    if (nguid_nz) {
+      // 6.1.3.1.1
+      result.page_length = 40;
+
+      // convert 128-bit number into hex string
+      char hex_string[33];
+      sprintf(hex_string, "%llx", identify_namespace_data.nguid[0]);
+      sprintf(&hex_string[16], "%llx", identify_namespace_data.nguid[1]);
```

Comment:
These format strings will strip leading zeroes, which probably isn't what you intended. IIRC `%08llx` is what you want here.